### PR TITLE
chore: Decouple `noirc_abi` from frontend by introducing `PrintableType`

### PR DESCRIPTION
**Pull request by phated**
# Description

## Problem\*

Towards #2238  

## Summary\*

This completes the decoupling of the `noirc_abi` from the `noirc_frontend` by introducing a new crate called `noirc_printable_type`. Right now, the `PrintableType` is similar to the `AbiType` but is a separate implementation to allow these to grow independently. For example, we probably want to print tuples but might not want to support them in the ABI, or the more concrete example of #2238.

In doing this work, I also noticed room for improvement in how `Display` was implemented for `PrintableType` to avoid round-tripping to serde_json. Here's the new output from running `nargo execute` on https://github.com/noir-lang/noir/blob/master/crates/nargo_cli/tests/execution_success/debug_logs/src/main.nr:
```bash
"i: {i}, j: {j}"
i: 0x01, j: 0x02
i: 0x01, j: 0x02
i: 0x01, j: 0x02
myStruct { y: 0x01, x: 0x02 }
randomstring0x010x01
i: 0x01, s: myStruct { y: 0x01, x: 0x02 }
0x01
[0x01, 0x02]
s: myStruct { y: 0x01, x: 0x02 }, foo: fooStruct { my_struct: myStruct { y: 0x02, x: 0x01 }, foo: 0x02 }
x: 0, y: 1
s1: myStruct { y: 0x01, x: 0x02 }, s2: myStruct { y: 0x02, x: 0x1e }
foo1: fooStruct { my_struct: myStruct { y: 0x01, x: 0x02 }, foo: 0x0f }, foo2: fooStruct { my_struct: myStruct { y: 0x02, x: 0x0f }, foo: 0x1e }
fooStruct { my_struct: myStruct { y: 0x01, x: 0x02 }, foo: 0x0f }
```

I'm happy to make any println formatting changes here if we don't like this output.

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: crates/noirc_frontend/src/ast/expression.rs [hunk 1]
```diff
@@ -382,8 +382,7 @@ pub enum FunctionReturnType {
 
 /// Describes the types of smart contract functions that are allowed.
 /// - All Noir programs in the non-contract context can be seen as `Secret`.
-#[derive(serde::Serialize, serde::Deserialize, Debug, Copy, Clone, PartialEq, Eq)]
-#[serde(rename_all = "snake_case")]
```
Comment (kevaundray) [inline crates/noirc_frontend/src/ast/expression.rs:386]: This seems like a breaking change, why was this needed?

Comment (phated) [reply crates/noirc_frontend/src/ast/expression.rs:386]: This wasn't used anywhere otherwise it wouldn't compile 🙃

Comment (phated) [reply crates/noirc_frontend/src/ast/expression.rs]: I removed it from this PR but will followup to remove serde from the frontend crate.

Comment (kevaundray) [reply crates/noirc_frontend/src/ast/expression.rs]: Oh thats suprising haha -- I thought it was being used for contract ABIs 🤔

Comment (phated) [reply crates/noirc_frontend/src/ast/expression.rs]: I think it's be translated somewhere else. I'll still do it in a follow up PR since it is unrelated to this code but was a nice cleanup to remove the dep

File: crates/noirc_abi/src/lib.rs [hunk 2]
```diff
@@ -97,6 +116,60 @@ pub enum Sign {
...
+            Type::Error => unreachable!(),
+            Type::Unit => unreachable!(),
+            Type::Constant(_) => unreachable!(),
+            Type::Struct(def, ref args) => {
+                let struct_type = def.borrow();
+                let fields = struct_type.get_fields(args);
+                let fields = vecmap(fields, |(name, typ)| (name, Self::from_type(&typ)));
+                Self::Struct { fields, name: struct_type.name.to_string() }
+            }
+            Type::Tuple(_) => todo!("as_abi_type not yet implemented for tuple types"),
```
Comment (vezenovm) [inline crates/noirc_abi/src/lib.rs:163]: Nit: as_abi_type is no longer the name of the function where this panic occurs

File: crates/noirc_abi/src/lib.rs [hunk 3]
```diff
@@ -97,6 +116,60 @@ pub enum Sign {
 }
 
 impl AbiType {
+    // TODO: Add `Context` argument for resolving fully qualified struct paths
```
Comment (vezenovm) [inline crates/noirc_abi/src/lib.rs:119]: Can you link issue here before we merge?

File: crates/noirc_abi/src/lib.rs [hunk 4]
```diff
@@ -97,6 +116,60 @@ pub enum Sign {
...
+    pub fn from_type(typ: &Type) -> Self {
+        // Note; use strict_eq instead of partial_eq when comparing field types
+        // in this method, you most likely want to distinguish between public and private
```
Comment (vezenovm) [inline crates/noirc_abi/src/lib.rs:122]: Why most likely? Do you mean the users who are consuming the ABI such as Aztec?

Comment (vezenovm) [reply crates/noirc_abi/src/lib.rs:122]: I see now this is the note that was already on top of `as_abi_type()`

File: crates/noirc_frontend/src/monomorphization/mod.rs [hunk 5]
```diff
@@ -875,15 +875,15 @@ impl<'interner> Monomorphizer<'interner> {
...
         if let HirType::Array(size, _) = typ {
             if let HirType::NotConstant = **size {
                 unreachable!("println does not support slices. Convert the slice to an array before passing it to println");
             }
         }
-        let abi_type = typ.as_abi_type();
+        let printable_type: PrintableType = typ.into();
         let abi_as_string =
-            serde_json::to_string(&abi_type).expect("ICE: expected Abi type to serialize");
+            serde_json::to_string(&printable_type).expect("ICE: expected Abi type to serialize");
```
Comment (vezenovm) [inline crates/noirc_frontend/src/monomorphization/mod.rs:886]: Nit: expected `PrintableType` to be serialized

File: crates/noirc_printable_type/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    }
+}
+
+impl<'a> std::fmt::Display for PrintableValueDisplay<'a> {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match (&self.value, &self.typ) {
+            (
+                PrintableValue::Field(f),
+                PrintableType::Field
+                // TODO: We should print the sign for these and probably print normal integers instead of field strings
```
Comment (vezenovm) [inline crates/noirc_printable_type/src/lib.rs:80]: Could we make an issue for this TODO? Could be a good first issue for a new contributor

File: crates/noirc_printable_type/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+/// and those TomlTypes will be mapped to Value
+#[derive(Debug, Clone, Serialize, PartialEq)]
+pub enum PrintableValue {
+    Field(FieldElement),
+    String(String),
+    Vec(Vec<PrintableValue>),
+    Struct(BTreeMap<String, PrintableValue>),
+}
+
+/// In order to display an `PrintableValue` we need an `PrintableType` to accurately
```
Comment (vezenovm) [inline crates/noirc_printable_type/src/lib.rs:60]:
```suggestion
/// In order to display a `PrintableValue` we need a `PrintableType` to accurately
```

File: crates/noirc_printable_type/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+                struct_map.insert(field_key.to_owned(), field_value);
+            }
+
+            PrintableValue::Struct(struct_map)
+        }
+    }
+}
+
+// TODO: Figure out a better API for this to avoid exporting the function
+pub fn decode_string_value(field_elements: &[FieldElement]) -> String {
```
Comment (vezenovm) [inline crates/noirc_printable_type/src/lib.rs:189]: I believe this is a repeat as we still have a `decode_string_value` in `noirc_abi`? Could we switch to using one or the other? Or is this part of the TODO at the top of this method?

File: crates/noirc_printable_type/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+    let mut trimmed_field = field.to_hex().trim_start_matches('0').to_owned();
+    if trimmed_field.len() % 2 != 0 {
+        trimmed_field = "0".to_owned() + &trimmed_field;
+    }
+    "0x".to_owned() + &trimmed_field
+}
+
+// TODO: Figure out a better API for this to avoid exporting the function
+/// Assumes that `field_iterator` contains enough [FieldElement] in order to decode the [PrintableType]
+pub fn decode_value(
```
Comment (vezenovm) [inline crates/noirc_printable_type/src/lib.rs:147]: Could we instead have a `from_values` method of some kind on `PrintableValueDisplay` that takes in a `PrintableType` and list of values? The `from` method then uses `decode_value` to construct a `PrintableValue`.

File: crates/noirc_abi/src/lib.rs (hunk 3 above)
Comment (phated) [reply crates/noirc_abi/src/lib.rs:119]: This is solved and removed in #2374 so no need to track it

File: crates/noirc_abi/src/lib.rs (hunk 2 above)
Comment (phated) [reply crates/noirc_abi/src/lib.rs:163]:
```suggestion
            Type::Tuple(_) => todo!("AbiType::from_type not yet implemented for tuple types"),
```

File: crates/noirc_printable_type/src/lib.rs (hunk 8 above)
Comment (phated) [reply crates/noirc_printable_type/src/lib.rs:189]: This crate and `noirc_abi` cannot use functions from each other so they don't cause a circular dependency, so this function and the function above are currently duplicated in each crate but they probably shouldn't be public at and instead should be implemented as `From` or serde.

File: crates/noirc_printable_type/src/lib.rs (hunk 6 above)
Comment (phated) [reply crates/noirc_printable_type/src/lib.rs:80]:
```suggestion
                // TODO(#2401): We should print the sign for these and probably print normal integers instead of field strings
```

File: crates/noirc_printable_type/src/lib.rs (hunk 9 above)
Comment (phated) [reply crates/noirc_printable_type/src/lib.rs]: @vezenovm I just did a big refactor so `PrintableValueDisplay` could implement `TryFrom<&[Vec<Value>]>`. Please take a look.

File: crates/nargo/src/ops/foreign_calls.rs [hunk 10]
```diff
@@ -78,90 +77,9 @@ impl ForeignCall {
...
-            foreign_call_inputs.split_last().ok_or(ForeignCallError::MissingForeignCallInputs)?;
-
-        let output_string = if is_fmt_str[0].to_field().is_one() {
-            convert_fmt_string_inputs(foreign_call_inputs)?
-        } else {
-            convert_string_inputs(foreign_call_inputs)?
-        };
-        println!("{output_string}");
+    fn execute_println(foreign_call_inputs: &[Vec<Value>]) -> Result<(), NargoError> {
+        let display_values: PrintableValueDisplay = foreign_call_inputs.try_into()?;
```
Comment (vezenovm) [inline crates/nargo/src/ops/foreign_calls.rs:81]: nice

File: crates/noirc_printable_type/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+
+/// In order to display a `PrintableValue` we need a `PrintableType` to accurately
+/// convert the value into a human-readable format.
+pub enum PrintableValueDisplay {
+    Plain(PrintableValue, PrintableType),
+    FmtString(String, Vec<(PrintableValue, PrintableType)>),
+}
+
+#[derive(Debug, Error)]
+pub enum ForeignCallError {
```
Comment (vezenovm) [inline crates/noirc_printable_type/src/lib.rs:70]: I see why you moved these here. If we ever decide to have compiler builtin foreign calls not related to printing we can always bring back the error enum into nargo

**Review by vezenovm — APPROVED** [review]
